kf/bpf.go

```diff
@@ -1235,3 +1197,283 @@ func ValidatePath(filePath string, destination string) (string, error) {
+	}
+
+	progInfo, err := bpfProg.Info()
+	if err != nil {
+		return fmt.Errorf("%s: information of bpf program failed", b.Program.Name)
+	}
+
+	ok := false
+	b.ProgID, ok = progInfo.ID()
+	if !ok {
+		log.Warn().Msgf("Program ID fetch failed: %s", b.Program.Name)
+	}
+
+	// Initialise metric maps
+	if err := b.InitialiseMetricMaps(); err != nil {
+		return fmt.Errorf("initialising metric maps failed %#v", err)
+	}
+
+	// Update program map id
+	ss := strings.Split(b.Program.MapName, "/")
+	progMapName := ss[len(ss)-1]
+	progMap := prg.Maps[progMapName]
+	if progMap == nil {
+		log.Error().Msgf("program map fetch failed: %s", progMapName)
+		return fmt.Errorf("program map fetch failed: %s", progMapName)
+	}
+
+	progMapInfo, err := progMap.Info()
+	if err != nil {
+		return fmt.Errorf("unable to access pinned previous prog map %s ebpfmapinfo %v", b.MapNamePath, err)
+	}
+	b.ProgMapID, ok = progMapInfo.ID()
+	if !ok {
+		return fmt.Errorf("eBPF program %s failed to fetch map id", b.Program.Name)
+	}
+
+	// tc and xdp maps pinning
+	for k, v := range prg.Maps {
+		var mapFilename string
+		if b.Program.ProgType == models.TCType {
+			mapFilename = filepath.Join(b.hostConfig.BpfMapDefaultPath, models.TCMapPinPath, ifaceName, k)
+		} else {
+			mapFilename = filepath.Join(b.hostConfig.BpfMapDefaultPath, ifaceName, k)
+		}
+		// In case one of the program pins the map then other program will skip
+		if !fileExists(mapFilename) {
+			if err := v.Pin(mapFilename); err != nil {
+				return fmt.Errorf("BPF program %s map %s:failed to pin the map err - %#v", b.Program.Name, mapFilename, err)
+			}
+		}
+	}
+
+	// Link this program into previous program map
+	ebpfMap, err := ebpf.NewMapFromID(b.PrevProgMapID)
+	if err != nil {
+		return fmt.Errorf("unable to access pinned previous prog map %s %v", b.PrevMapNamePath, err)
+	}
+	defer ebpfMap.Close()
+
+	key := 0
+	fd := bpfProg.FD()
+	log.Info().Msgf("previous program map path %s FD %d", b.PrevMapNamePath, fd)
+	if err = ebpfMap.Update(unsafe.Pointer(&key), unsafe.Pointer(&fd), 0); err != nil {
+		return fmt.Errorf("unable to update prog next map %s %v", b.Program.MapName, err)
+	}
+	log.Info().Msgf("BPF program %s loaded on interface %s direction %s successfully", b.Program.Name, ifaceName, direction)
+	return nil
+}
+
+// InitialiseMetricMaps - This method initialises all the monitor maps
+func (b *BPF) InitialiseMetricMaps() error {
+	if b.ProgMapCollection == nil {
+		log.Warn().Msgf("prog is not loaded by l3afd")
+		return nil
+	}
+	for _, tmpMap := range b.Program.MonitorMaps {
+		tmpMetricsMap := b.ProgMapCollection.Maps[tmpMap.Name]
+		if tmpMetricsMap == nil {
+			log.Error().Msgf("%s map is not loaded", tmpMap.Name)
+			continue
+		}
+
+		var err error
+		log.Debug().Msgf("Program - %s map name %s key size %d value size %d\n", b.Program.Name, tmpMap.Name, tmpMetricsMap.KeySize(), tmpMetricsMap.ValueSize())
+		if tmpMetricsMap.KeySize() == 1 {
+			var k int8
+			switch tmpMetricsMap.ValueSize() {
```

Comment:
Can we move this into a function as it is repeated multiple times below. for other key sizes?

Reply:
Here we need to have 4 separate functions or all in same function, because different key size makes define key of different types